sdk/eidos_sdk/util/stream_collector.py

```diff
@@ -1,15 +1,36 @@
 from __future__ import annotations
 
-from typing import Optional
+import copy
+from collections import deque
+from typing import Optional, AsyncIterator, Deque, Callable
 
-from eidos_sdk.io.events import BaseStreamEvent, StringOutputEvent, ObjectOutputEvent, ErrorEvent
+from eidos_sdk.io.events import (
+    BaseStreamEvent,
+    StringOutputEvent,
+    ObjectOutputEvent,
+    ErrorEvent,
+    StreamEvent,
+    StartStreamContextEvent,
+    EndStreamContextEvent,
+)
 
 
-class StringStreamCollector:
-    contents: Optional[str]
+class StreamCollector(AsyncIterator[StreamEvent]):
```

Comment:
I turned StreamCollector into a pretty helpful wrapper class. It can still be used the old way to just collect as well. I was going to make this a separate class, but think it fits it's role well. We also handle exceptions by throwing them on the event stream and re-raising a related exception to force the consumer to handle or ignore it at the higher level.